server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -506,75 +505,56 @@ protected Symbol visitFunctionCall(FunctionCall node, ExpressionAnalysisContext
-            List<Symbol> operands = new ArrayList<>(whenClauses.size());
-            List<Symbol> results = new ArrayList<>(whenClauses.size());
-            Symbol caseOperand = convert(node.getOperand(), context);
+            var whenClauses = node.getWhenClauses();
+            var arguments = new ArrayList<Symbol>(whenClauses.size() * 2 + 2);
+
+            var defaultValue = node.getDefaultValue();
+            arguments.add(convert(BooleanLiteral.TRUE_LITERAL, context));
+
+            if (defaultValue == null) {
+                arguments.add(convert(NullLiteral.INSTANCE, context));
+            } else {
+                arguments.add(convert(defaultValue, context));
+            }
+
             for (WhenClause whenClause : whenClauses) {
-                Symbol whenOperand = convert(whenClause.getOperand(), context);
-                operands.add(allocateFunction(EqOperator.NAME, List.of(caseOperand, whenOperand), context));
-                results.add(convert(whenClause.getResult(), context));
+                arguments.add(buildCondition(node.getOperand(), whenClause.getOperand(), context));
+                arguments.add(whenClause.getResult().accept(this,context));
             }
-            ensureResultTypesMatch(results);
-            Expression defaultValue = node.getDefaultValue();
-            return createChain(
-                operands,
-                results,
-                defaultValue == null ? null : convert(defaultValue, context),
+            ensureResultTypesMatch(arguments);
+            return allocateFunction(CaseFunction.NAME, arguments, context);
+        }
+
+        private Symbol buildCondition(Expression nodeOperand,
+                                      Expression value,
+                                      ExpressionAnalysisContext context) {
+            return allocateFunction(
+                EqOperator.NAME,
+                List.of(convert(nodeOperand, context), convert(value, context)),
                 context
             );
         }
 
         @Override
         protected Symbol visitSearchedCaseExpression(SearchedCaseExpression node, ExpressionAnalysisContext context) {
-            List<WhenClause> whenClauses = node.getWhenClauses();
-            List<Symbol> operands = new ArrayList<>(whenClauses.size());
-            List<Symbol> results = new ArrayList<>(whenClauses.size());
-            for (WhenClause whenClause : whenClauses) {
-                operands.add(whenClause.getOperand().accept(innerAnalyzer, context));
-                results.add(whenClause.getResult().accept(innerAnalyzer, context));
+            var whenClauses = node.getWhenClauses();
+            var arguments = new ArrayList<Symbol>(whenClauses.size() * 2 + 2);
+
+            var defaultValue = node.getDefaultValue();
+            arguments.add(convert(BooleanLiteral.TRUE_LITERAL, context));
+
+            if (defaultValue == null) {
+                arguments.add(convert(NullLiteral.INSTANCE, context));
```

Comment:
See above, we've literal singletons, no need to convert the parser expressions to get them.